### ISSUE TITLE
[bug] fails to build with gcc 14

### ISSUE THREAD

**Issue by darix** [sparklemotion/nokogiri#3295]
openSUSE Tumbleweed is currently in the process of switching to gcc 14. Which brings yet more stricter enforcements.

```
Gem files will remain installed in /home/abuild/rpmbuild/BUILDROOT/rubygem-nokogiri-1.15.5-3.15.x86_64/usr/lib64/ruby/gems/3.3.0/gems/nokogiri-1.15.5 for inspection.
Results logged to /home/abuild/rpmbuild/BUILDROOT/rubygem-nokogiri-1.15.5-3.15.x86_64/usr/lib64/ruby/gems/3.3.0/extensions/x86_64-linux-gnu/3.3.0/nokogiri-1.15.5/gem_make.out
compiling gumbo.c
compiling html4_document.c
html4_document.c: In function ‘rb_html_document_s_read_io’:
html4_document.c:49:52: error: passing argument 2 of ‘xmlSetStructuredErrorFunc’ from incompatible pointer type [-Wincompatible-pointer-types]
   49 |   xmlSetStructuredErrorFunc((void *)rb_error_list, Nokogiri_error_array_pusher);
      |                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                    |
      |                                                    void (*)(void *, xmlError *) {aka void (*)(void *, struct _xmlError *)}
In file included from /usr/include/libxml/valid.h:15,
                 from /usr/include/libxml/parser.h:19,
                 from ./nokogiri.h:35,
                 from html4_document.c:1:
/usr/include/libxml/xmlerror.h:898:57: note: expected ‘xmlStructuredErrorFunc’ {aka ‘void (*)(void *, const struct _xmlError *)’} but argument is of type ‘void (*)(void *, xmlError *)’ {aka ‘void (*)(void *, struct _xmlError *)’}
  898 |                                  xmlStructuredErrorFunc handler);
      |                                  ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
html4_document.c: In function ‘rb_html_document_s_read_memory’:
html4_document.c:109:52: error: passing argument 2 of ‘xmlSetStructuredErrorFunc’ from incompatible pointer type [-Wincompatible-pointer-types]
  109 |   xmlSetStructuredErrorFunc((void *)rb_error_list, Nokogiri_error_array_pusher);
      |                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                    |
      |                                                    void (*)(void *, xmlError *) {aka void (*)(void *, struct _xmlError *)}
/usr/include/libxml/xmlerror.h:898:57: note: expected ‘xmlStructuredErrorFunc’ {aka ‘void (*)(void *, const struct _xmlError *)’} but argument is of type ‘void (*)(void *, xmlError *)’ {aka ‘void (*)(void *, struct _xmlError *)’}
  898 |                                  xmlStructuredErrorFunc handler);
      |                                  ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
```

**Comment by darix** [comment]
on older compilers you can get the same behavior with

```
-Werror=implicit-int -Werror=implicit-function-declaration -Werror=declaration-missing-parameter-type -Werror=return-mismatch -Werror=int-conversion -Werror=incompatible-pointer-types
```

**Comment by flavorjones** [comment]
@darix Thanks for opening this issue.

The issue you're seeing was because upstream libxml2 changed its function signature, and was fixed in Nokogiri in fa3336e1 which first appeared in v1.16.0.

Looks like you're compiling Nokogiri v1.15.5, however the current supported version is v1.16.6 -- I'm not supporting v1.15 at this time. Hopefully you'll see this is fixed when you update Nokogiri.
